unitypack/export.py

```diff
@@ -0,0 +1,212 @@
+
+
+def extract_indicies(mesh):
+	indices = []
+	triangles = []
+	for sub in mesh.submeshes:
+		sub_indices = []
+		sub_triangles = []
+		buf = BinaryReader(BytesIO(mesh.index_buffer))
+		buf.seek(sub.first_byte)
+		for i in range(0, sub.index_count):
+			sub_indices.append(buf.read_uint16())
+		if sub.topology == 0:
+			sub_triangles.extend(sub_indices)
+		else:
+			raise NotImplementedError("(%s) topologies are not supported" % (mesh.name))
+
+		indices.append(sub_indices)
+		triangles.append(sub_triangles)
+
+	return {
+		"indices": indices,
+		"triangles": triangles
+	}
+
+
+def extract_vertices(mesh):
+	vertices = []
+	normals = []
+	colors = []
+	uv1 = []
+	uv2 = []
+	uv3 = []
+	uv4 = []
+	tangents = []
+	# unity 5+ has 8 channels (6 otherwise)
+	v5_channel_count = 8
+	v5_channels = False
+
+	buf = BinaryReader(BytesIO(mesh.vertex_data.data))
+	channels = mesh.vertex_data.channels
+	# actual streams attribute 'm_Streams' may only exist in unity 4,
+	# use of channel data alone seems to be sufficient
+	streams = get_num_streams(channels)
+	channel_count = len(channels)
+	if channel_count == v5_channel_count:
+		v5_channels = True
+
+	for s in range(0, streams):
+		for i in range(0, mesh.vertex_data.vertex_count):
+			for j in range(0, channel_count):
+				ch = None
+				if channel_count > 0:
+					ch = channels[j]
+					# format == 1, use half-floats (16 bit)
+					if ch["format"] == 1:
+						raise NotImplementedError("(%s) 16 bit floats are not supported" % (mesh.name))
+				# read the appropriate vertex value into the correct list
+				if ch and ch["dimension"] > 0 and ch["stream"] == s:
+					if j == 0:
+						vertices.append(Vec3().read(buf))
+					elif j == 1:
+						normals.append(Vec3().read(buf))
+					elif j == 2:
+						colors.append(Vec4().read_color(buf))
+					elif j == 3:
+						uv1.append(Vec2().read(buf))
+					elif j == 4:
+						uv2.append(Vec2().read(buf))
+					elif j == 5:
+						if v5_channels:
+							uv3.append(Vec2().read(buf))
+						else:
+							tangents.append(Vec4().read(buf))
+					elif j == 6: # for unity 5+
+						uv4.append(Vec2().read(buf))
+					elif j == 7: # for unity 5+
+						tangents.append(Vec4().read(buf))
+		# TODO investigate possible alignment here, after each stream
+
+	return {
+		"vertices": vertices,
+		"normals": normals,
+		"colors" : colors,
+		"uv1" : uv1,
+		"uv2" : uv2,
+		"uv3" : uv3,
+		"uv4" : uv4,
+		"tangents" : tangents
+	}
+
+
+class OBJMesh:
+	def __init__(self, mesh):
+		self.mesh = mesh
+		if self.mesh.mesh_compression:
+			# TODO handle compressed meshes
+			raise NotImplementedError("(%s) compressed meshes are not supported" % (mesh.name))
+		else:
+			self.indicies = extract_indicies(mesh)
+			self.vertices = extract_vertices(mesh)
+
+	def vec2_str(self, prefix, vec):
+		return "%s %s %s\n" % (prefix, vec.x, 1 - vec.y)
+
+	def vec3_str(self, prefix, vec):
+		return "%s %s %s %s\n" % (prefix, -vec.x, vec.y, vec.z)
+
+	def face_str(self, indices, vts_len, vns_len):
```

Comment:
Some to think of it, you can just implement `vecX_str` on the class itself, which will let you do `str(vec)`. Much simpler.